subsys/bluetooth/host/gatt.c

```diff
@@ -2322,14 +2323,17 @@ static void gatt_find_type_rsp(struct bt_conn *conn, u8_t err,
 		       end_handle);
 
 		if (params->type == BT_GATT_DISCOVER_PRIMARY) {
-			attr.uuid = BT_UUID_GATT_PRIMARY;
+			memcpy(&uuid_svc, BT_UUID_GATT_PRIMARY,
```

Comment:
I am just not sure if this is optimal way to do it - we are creating object in RAM, initializing it, and after that copying data using memcpy... and then the object is not valid any more. Technically it is correct, but we can optimise it:
```
uuid_svc.uuid.type = BT_UUID_TYPE_16;
if (params->type == BT_GATT_DISCOVER_PRIMARY) {
    uuid_svc.val = BT_UUID_16(BT_UUID_GATT_PRIMARY)->val;
} else {
    uuid_svc.val = BT_UUID_16(BT_UUID_GATT_SECONDARY)->val
}
```

With a hope that compiler would optimise ```BT_UUID_16(BT_UUID_GATT_PRIMARY)->val``` to one constant.

Reply:
Thanks. Update to this. I agree a memcpy on 2 bytes is a bit much

Reply:
> With a hope that compiler would optimise BT_UUID_16(BT_UUID_GATT_PRIMARY)->val to one constant.

I realise we can't do it for the release, but the need for this kind of indirection is indicative of a suboptimal or incomplete API definition. We should probably have defines for the literal integers values so we can use them directly. Any chance someone could check that the compiler really does optimize the short-term solution and that we're not left with an extra 3-byte struct and a pointer to that struct, just to get the value stored in 2 bytes of the struct?

Reply:
It looks good (with size optimisation):
```
		if (params->type == BT_GATT_DISCOVER_PRIMARY) {
   19bd0:	f896 3028 	ldrb.w	r3, [r6, #40]	; 0x28
		end_handle = sys_le16_to_cpu(rsp->list[i].end_handle);
   19bd4:	f8b1 9002 	ldrh.w	r9, [r1, #2]
		attr.handle = start_handle;
   19bd8:	f8ad 201c 	strh.w	r2, [sp, #28]
			uuid_svc.val = BT_UUID_16(BT_UUID_GATT_SECONDARY)->val;
   19bdc:	2b00      	cmp	r3, #0
   19bde:	f642 0301 	movw	r3, #10241	; 0x2801
   19be2:	bf08      	it	eq
   19be4:	f44f 5320 	moveq.w	r3, #10240	; 0x2800
   19be8:	f8ad 3002 	strh.w	r3, [sp, #2]
```

You can see at 19bde and 19be4 that simple constant values are used.

Reply:
@rakons Thank you for that analysis.

@jhedberg I thought about adding defines for the UUID values. There is a lot of them though.
But with those values I could have used `BT_UUID_INIT_16` instead. I think we could atleast have for the most standard ones, like characterstic and service. But didn't want to do it for this fix.